osbenchmark/synthetic_data_generator/synthetic_data_generator.py

```diff
@@ -6,98 +6,148 @@
-        console.println(f"User wants to generate [{sdg_config.total_size_gb}]GB but current host does not have enough available disk storage")
-        sys.exit(1)
+from osbenchmark.synthetic_data_generator.types import GB_TO_BYTES
+from osbenchmark.synthetic_data_generator.helpers import write_chunk, get_generation_settings, setup_custom_tqdm_formatting, check_for_exsiting_files
+from osbenchmark.synthetic_data_generator.strategies import DataGenerationStrategy
+from osbenchmark.synthetic_data_generator.types import SyntheticDataGeneratorMetadata
+
+class SyntheticDataGenerator:
+    def __init__(self, sdg_metadata: SyntheticDataGeneratorMetadata, sdg_config: dict, strategy: DataGenerationStrategy) -> None:
+        self.sdg_metadata = sdg_metadata
+        self.sdg_config = sdg_config
+        self.strategy = strategy
+
+        self.logger = logging.getLogger(__name__)
+
+    def generate_seeds_for_workers(self, regenerate=False):
+        # This adds latency so might consider deprecating this
+        client = get_client()
+        workers = client.scheduler_info()['workers']
+
+        seeds = []
+        for worker_id in workers.keys():
+            hash_object = hashlib.md5(worker_id.encode())
+
+            if regenerate:
+                # Add current timestamp to each hash to improve uniqueness
+                timestamp = str(time.time()).encode()
+                hash_object.update(timestamp)
+
+            hash_hex = hash_object.hexdigest()
+
+            seed = int(hash_hex[:8], 16)
+            seeds.append(seed)
+
+        return seeds
+
+    def generate_test_document(self):
+        return self.strategy.generate_test_document()
+
+    def generate_dataset(self):
+        """
+        Core logic in generating synthetic data. Can use different strategies
+        """
+        generation_settings: dict = get_generation_settings(self.sdg_config)
+        max_file_size_bytes: int = (generation_settings.get('max_file_size_gb') or 0) * GB_TO_BYTES
+        total_size_bytes: int = self.sdg_metadata.total_size_gb * GB_TO_BYTES
+        docs_per_chunk: int = generation_settings.get('docs_per_chunk')
+
+        avg_document_size = self.strategy.calculate_avg_doc_size()
+
+        current_size = 0
+        docs_written = 0
+        file_counter = 0
+
+        generated_dataset_details = []
+
+        check_for_exsiting_files(self.sdg_metadata.output_path, self.sdg_metadata.index_name)
+
+        workers: int = generation_settings.get("workers")
+        dask_client = Client(n_workers=workers, threads_per_worker=1)  # We keep it to 1 thread because generating random data is CPU intensive
+        self.logger.info("Number of workers to use: [%s]", workers)
+
+        console.println(f"[NOTE] ✨ Dashboard link to monitor processes and task streams: [{dask_client.dashboard_link}]")
+        console.println("[NOTE] ✨ For users who are running generation on a virtual machine, consider tunneling to localhost to view dashboard.")
```

Comment:
More instructions on how to do this will be helpful.

Reply:
Wanted to link it in documentation but can add it in temporarily.